controller/backing_image_controller.go

```diff
@@ -310,22 +345,530 @@ func (bic *BackingImageController) syncBackingImage(key string) (err error) {
+	bic.v2CopyBackoff.Next(firstV2CopyDiskUUID, time.Now())
+	return nil
+}
+
+func (bic *BackingImageController) getFileDownloadAddress(bi *longhorn.BackingImage, diskUUID string) (string, error) {
+	bimName := types.GetBackingImageManagerName(bic.bimImageName, diskUUID)
+	bim, err := bic.ds.GetBackingImageManager(bimName)
+	if err != nil {
+		return "", err
+	}
+	bimCli, err := engineapi.NewBackingImageManagerClient(bim)
+	if err != nil {
+		return "", err
+	}
+	// get the file download address of sync server in the backing image manager
+	filePath, addr, err := bimCli.PrepareDownload(bi.Name, bi.Status.UUID)
+	if err != nil {
+		return "", err
+	}
+
+	fileDownloadAddress := fmt.Sprintf("http://%s/v1/files/%s/download", addr, url.PathEscape(filePath))
+	return fileDownloadAddress, nil
+}
+
+func (bic *BackingImageController) deleteInvalidV2Copy(bi *longhorn.BackingImage) (err error) {
+	log := getLoggerForBackingImage(bic.logger, bi)
+
+	if bi.Status.V2FirstCopyStatus != longhorn.BackingImageStateReady {
+		return nil
+	}
+
+	v2CopiesUnknown := []string{}
+	v2CopiesFailed := []string{}
+	missingSpecV2Copies := []string{}
+	hasReadyV2BackingImage := false
+
+	for v2DiskUUID, copyStatus := range bi.Status.DiskFileStatusMap {
+		// only handle v2 copy in this function
+		if types.IsDataEngineV1(copyStatus.DataEngine) {
+			continue
+		}
+
+		// Delete the v2 copy if the spec record is removed
+		if _, exists := bi.Spec.DiskFileSpecMap[v2DiskUUID]; !exists {
+			missingSpecV2Copies = append(missingSpecV2Copies, v2DiskUUID)
+			continue
+		}
+
+		if copyStatus.State == longhorn.BackingImageStateReady {
+			hasReadyV2BackingImage = true
+			continue
+		}
+
+		if copyStatus.State == longhorn.BackingImageStateFailed {
+			v2CopiesFailed = append(v2CopiesFailed, v2DiskUUID)
+			continue
+		}
+		if copyStatus.State == longhorn.BackingImageStateUnknown {
+			v2CopiesUnknown = append(v2CopiesUnknown, v2DiskUUID)
+			continue
+		}
+	}
+
+	for _, v2DiskUUID := range missingSpecV2Copies {
+		node, _, err := bic.ds.GetReadyDiskNodeRO(v2DiskUUID)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get the ready disk node for disk %v", v2DiskUUID)
+		}
+
+		engineClientProxy, err := bic.getEngineClientProxy(node.Name, longhorn.DataEngineTypeV2, log)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get the engine client proxy for node %v", node.Name)
+		}
+
+		log.Infof("Deleting the missing spec v2 copy on disk %v", v2DiskUUID)
+		err = engineClientProxy.SPDKBackingImageDelete(bi.Name, v2DiskUUID)
+		engineClientProxy.Close()
+		if err != nil {
+			return errors.Wrapf(err, "failed to delete the v2 backing image on diskUUID %v", v2DiskUUID)
+		}
+
+		delete(bi.Status.DiskFileStatusMap, v2DiskUUID)
+		bic.v2CopyBackoff.DeleteEntry(v2DiskUUID)
+		bic.eventRecorder.Eventf(bi, corev1.EventTypeNormal, constant.EventReasonDelete, "Deleted failed v2 backing image %v on disk %v on node %v", bi.Name, v2DiskUUID, node.Name)
+	}
```

Comment:
nit: abstract it to `cleanupMissingSpecV2Copies()`

Reply:
fixed.